### PR TITLE
Accelerate lp inference: Fetch node embs of multiple batches in one pull

### PR DESCRIPTION
**Pull request by isratnisa**
Resolves Issue https://github.com/awslabs/graphstorm/issues/169

To optimize network bandwidth usage, we concatenate node lists from multiple batches and retrieve their embeddings with a single remote pull. Heuristically, we have found using a batch size of 100K results in efficient network utilization and reduces the end-to-end inference pipeline from 45 to 21 minutes compared to using a batch size of 1024 on ogbn-papers100M dataset. 

### Reason not to merge this PR:

We have increased the default argument `eval_batch_size` from 1K to 10K in this PR https://github.com/awslabs/graphstorm/pull/66 , which reduced the time for full inference pipeline by 2x (45 mins to 19 min paprs100M dataset). This is primarily because, when `eval_batch_size` 10K is used network achieves sufficient usage compare to using 1K batch size (tested on g4dn.12xlarge). Also, GPUs achieve better utilization.

This PR fuses batches to create a long batch size of 100K length. Compare to 1K batch size, it provides 2x speedup (45 mins to 21 mins). But since we are already using default batch size of 10K which is providing enough network bandwidth saturation, this PR is not effective anymore.

### CODE REVIEW

File: python/graphstorm/model/lp_gnn.py [hunk 1]
```diff
@@ -107,6 +109,100 @@ def forward(self, blocks, pos_graph,
...
+        will be used to fetch embeddings in get_embs() function
+
+    Return
+    ------
+    node_list: Dict of tuple of tensors
+        Dictionary of edge type to concatenated node list stored in
+        a tuple
+    """
+    node_list = {}
+    canonical_etype = prev_canonical_etype = list(pos_neg_tuple.keys())[0]
```
Comment (classicsong) [inline python/graphstorm/model/lp_gnn.py:185]: Why we need canonical_etype and prev_canonical_etype? They are pointing to the same item.

File: python/graphstorm/model/lp_gnn.py [hunk 2]
```diff
@@ -107,6 +109,100 @@ def forward(self, blocks, pos_graph,
...
+    pos_src, neg_src, pos_dst, neg_dst = pos_neg_tuple[prev_canonical_etype]
+
+    for _ in range(num_batch_to_cat):
+        pos_neg_tuple_next, _ = next(loader, (None, None))
+        if pos_neg_tuple_next is not None:
+            canonical_etype = list(pos_neg_tuple_next.keys())[0]
```
Comment (classicsong) [inline python/graphstorm/model/lp_gnn.py:191]: I don't think you need canonical_etype = in L185

File: python/graphstorm/model/lp_gnn.py [hunk 3]
```diff
@@ -107,6 +109,100 @@ def forward(self, blocks, pos_graph,
...
+            if canonical_etype != prev_canonical_etype:
+                node_list[prev_canonical_etype] = (pos_src, neg_src, pos_dst, neg_dst)
```
Comment (classicsong) [inline python/graphstorm/model/lp_gnn.py:193]: Can you add some docstring telling what you are doing here?

File: python/graphstorm/model/lp_gnn.py [hunk 4]
```diff
@@ -107,6 +109,100 @@ def forward(self, blocks, pos_graph,
...
+        Dictionary of edge type to concatenated node list stored in
+        a tuple
+    """
+    node_list = {}
+    canonical_etype = prev_canonical_etype = list(pos_neg_tuple.keys())[0]
+    pos_src, neg_src, pos_dst, neg_dst = pos_neg_tuple[prev_canonical_etype]
+
+    for _ in range(num_batch_to_cat):
+        pos_neg_tuple_next, _ = next(loader, (None, None))
+        if pos_neg_tuple_next is not None:
```
Comment (classicsong) [inline python/graphstorm/model/lp_gnn.py:190]: what if pos_neg_tuple_next is None?

File: python/graphstorm/model/lp_gnn.py [hunk 5]
```diff
@@ -134,17 +230,47 @@ def lp_mini_batch_predict(model, emb, loader, device):
...
+                # Split the concatenated batch back into orginal batch size to avoid GPU OOM
+                for _ in range(num_batch_to_cat):
+                    batch_emb[etype] = (pos_src_emb[p_st: p_st + ps_size],
+                        neg_src_emb[ns_st: ns_st + ns_size] if neg_src_emb is not None else None,
+                        pos_dst_emb[p_st: p_st + ps_size],
+                        neg_dst_emb[ns_st: ns_st + ns_size] if neg_dst_emb is not None else None)
+
+                    # Compute score nd ranking of each batch
+                    score = decoder.calc_test_scores(batch_emb, neg_sample_type, device)
+                    for canonical_etype, s in score.items():
```
Comment (classicsong) [inline python/graphstorm/model/lp_gnn.py:262]:
```suggestion
                    for canonical_etype, (pos_score, neg_score) in score.items():
```

File: python/graphstorm/model/lp_gnn.py [hunk 6]
```diff
@@ -107,6 +109,100 @@ def forward(self, blocks, pos_graph,
...
+            neg_dst_emb = emb[vtype][neg_dst.reshape(-1,)]
+            neg_dst_emb = neg_dst_emb.reshape(neg_dst.shape[0], neg_dst.shape[1], -1)
+        elif neg_sample_type == BUILTIN_LP_JOINT_NEG_SAMPLER:
+            neg_dst_emb = emb[vtype][neg_dst]
+        else:
+            assert False, f"Unknow negative sample type {neg_sample_type}"
+    return (pos_src_emb, neg_src_emb, pos_dst_emb, neg_dst_emb)
+
+
+def cat_nodelist_of_multiple_batches(pos_neg_tuple, loader, num_batch_to_cat):
```
Comment (classicsong) [inline python/graphstorm/model/lp_gnn.py:159]: Can you add unit tests for cat_nodelist_of_multiple_batches?

File: python/graphstorm/model/edge_decoder.py [hunk 7]
```diff
@@ -332,23 +325,20 @@ def calc_test_scores(self, emb, pos_neg_tuple, neg_sample_type, device):
...
-        assert isinstance(pos_neg_tuple, dict) and len(pos_neg_tuple) == 1, \
-            "DotDecoder is only applicable to link prediction task with " \
-            "single target training edge type"
-        canonical_etype = list(pos_neg_tuple.keys())[0]
-        pos_src, neg_src, pos_dst, neg_dst = pos_neg_tuple[canonical_etype]
-        utype, _, vtype = canonical_etype
-        pos_src_emb = emb[utype][pos_src].to(device)
-        pos_dst_emb = emb[vtype][pos_dst].to(device)
+        assert isinstance(batch_embs, dict) and len(batch_embs) == 1, \
+        "DotDecoder is only applicable to link prediction task with " \
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply python/graphstorm/model/edge_decoder.py]: Done

File: python/graphstorm/model/edge_decoder.py [hunk 8]
```diff
@@ -304,20 +304,13 @@ def forward(self, g, h):    # pylint: disable=arguments-differ
...
-            Node embeddings.
-        pos_neg_tuple: dict of tuple
-            Positive and negative edges stored in a tuple:
-            tuple(positive source, negative source,
-            postive destination, negatve destination).
-            The positive edges: (positive source, positive desitnation)
-            The negative edges: (positive source, negative desitnation) and
-                                (negative source, positive desitnation)
+        batch_embs: dict of tuple
+            Node embeddings stored in a tuple for an edge type
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply python/graphstorm/model/edge_decoder.py]: Added description

**Comment by isratnisa** [comment]
We have increased the default argument eval_batch_size from 1K to 10K in this PR https://github.com/awslabs/graphstorm/pull/66 , which reduced the time for full inference pipeline by 2x (45 mins to 19 min paprs100M dataset). This is primarily because, when eval_batch_size 10K is used network achieves sufficient usage compare to using 1K batch size (tested on g4dn.12xlarge). Also, GPUs achieve better utilization.

This PR fuses batches to create a long batch size of 100K length. Compare to 1K batch size, it provides 2x speedup (45 mins to 21 mins). But since we are already using default batch size of 10K which is providing enough network bandwidth saturation, this PR is not effective anymore.

**Pull request closed without merging**